samza-core/src/main/java/org/apache/samza/util/ShutdownUtil.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.util;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Shutdown related utils
+ */
+public class ShutdownUtil {
+  private static final Logger LOG = LoggerFactory.getLogger(ShutdownUtil.class);
+
+  /**
+   * A helper to facilitate shutting down a set of resources in parallel to enforce a bounded shutdown time.
+   * The helper function instantiates an {@link ExecutorService} to execute the "shutdown function", and will
+   * await the termination for given timeout. If shutdown remains unfinished in the end, the whole thread dump
+   * will be printed to help debugging.
+   *
+   * The shutdown is performed with best-effort. Depending on the implementation of the shutdown function, resource
+   * leak might be possible.
+   *
+   * @param shutdownFunction the shutdown function will be provided with an ExecutorService and is supposed to
+   *                         submit all the shutdown tasks via the given executor
+   * @param message message that will show in the thread name and the thread dump
+   * @param timeoutMs timeout in ms
+   */
+  public static void boundedShutdown(Function<ExecutorService, Void> shutdownFunction, String message, long timeoutMs) {
```

Comment:
1. Does the caller need access to the executor service? Can they caller provide a `List<Runnable> tasks`.
2. [Optional] Can this be a general method unrelated to shutdown? Seems like it essentially takes a List<Runnable/Callable>, executes them concurrently and tries to cancel them and return after waiting for a bounded amount of time.

Reply:
+1

Reply:
additionally, can return the status as boolean if all `Runnable`s complete.

Reply:
Discussed with Prateek offline. Totally makes sense to change the signature to just take a list of Runnable. Will also return the status as boolean.

In terms of making it a general method for concurrent execution, that would probably go too far. I still think this is more specific for shutdown because of the fact that Samza requires the system consumer to shutdown within a timeout during rebalance. In general, people should simply use an ExecutorService for executing concurrent tasks in a normal way.  So I prefer keeping this as a shutdown util for now to discourage people to use it for general purpose.